atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CqlExecutorImpl.java

```diff
@@ -81,23 +89,43 @@
         CqlPreparedResult preparedResult = queryExecutor.prepare(queryBytes, rowsAscending.get(0), Compression.NONE);
         int queryId = preparedResult.getItemId();
 
-        List<CellWithTimestamp> result = Lists.newArrayList();
         Iterator<byte[]> rows = rowsAscending.iterator();
-        while (result.size() < limit && rows.hasNext()) {
-            // We can parallelise this, but would need to collect it carefully to make sure we satisfy the guarantees:
-            // 1. Results are ordered
-            // 2. There are at most <limit> results
-            // 3. We return the first <limit> results
-            // We can ensure this by not applying the limit until after collection, and grouping by row
-            // I think this is ensured later on in the code path?
-            CqlResult cqlResult = queryExecutor.executePrepared(queryId,
-                    ImmutableList.of(ByteBuffer.wrap(rows.next())));
-            result.addAll(CqlExecutorImpl.getCells(CqlExecutorImpl::getCellFromRow, cqlResult));
+        List<CellWithTimestamp> result = Lists.newArrayList();
+        List<Future<CqlResult>> futures = Lists.newArrayList();
+
+        for (int i = 0; i < 16 && rows.hasNext(); i++) {
+            futures.add(submitAndGetFuture(queryId, rows.next()));
+        }
+
+        try {
+            for (int i = 0; i < futures.size(); i++) {
+                Future<CqlResult> future = futures.get(i);
+                CqlResult cqlResult = future.get();
+                result.addAll(CqlExecutorImpl.getCells(CqlExecutorImpl::getCellFromRow, cqlResult));
+
+                if (result.size() > limit) {
+                    break;
```

Comment:
We should cancel any remaining Futures before returning.  This way we won't execute queries for which we're never going to look at the results.  Need to check if the `queryExecutor` or some other part of the future checks for cancellation.

Reply:
Added the cancellations (pushing in a few mins). The `queryExecutor` pretty much delegates to `clientPool.runWithRetryOnHost`, so this may be interrupted - are there any concerns about that?

Reply:
Concern was if nothing checked it then work would still happen.  But I imagine it happens before the task starts running and when the query is sent and received over the network, which is probably enough.